File: scripts/R/0_update_param_site.R

rm(list = ls(all.names = TRUE))

library(httr)  # Useful tools for working with HTTP organised by HTTP verbs
# library(data.table)  # Fast operations on large data frames

time_start <- Sys.time()  # Start the timer


end_point <- "https://aquarius.orc.govt.nz/AQUARIUS/Publish/v2"

# =========================================================================
# --- 'GetLocationDescriptionList': plate numbers (ID) <-> Names (Site) ---
# =========================================================================
desc_r <- GET(
  paste0(end_point, "/GetLocationDescriptionList"),
  authenticate("api-read", "PR98U3SKOczINoPHo7WM")
)
stop_for_status(desc_r, cat("Check the URL for the requested data!\n"))
desc <- content(desc_r)$LocationDescriptions
# plate_df <- rbindlist(lapply(desc, "[", c("Identifier", "Name")))[, unique(.SD)]
# setnames(plate_df, old = c("Identifier", "Name"), new = c("ID", "Site"))
select_col_1 <- c("Identifier", "Name")
new_name_1 <- c("ID", "Site")
plate_df <-
  lapply(desc, "[", name = select_col_1) |>
  do.call(rbind, args = _) |>
  unlist() |>
  matrix(ncol = length(select_col_1)) |>
  unique() |>
  as.data.frame() |>
  `colnames<-`(new_name_1)


# ============================================
# --- 'GetParameterList': Unit_id <-> Unit ---
# ============================================
param_r <- GET(
  paste0(end_point, "/GetParameterList"),
  authenticate("api-read", "PR98U3SKOczINoPHo7WM")
)
stop_for_status(param_r, cat("Check the URL for the requested data!\n"))
param <- content(param_r)$Parameters
# param_df <- rbindlist(lapply(param, "[", c("Identifier", "UnitIdentifier")))[, unique(.SD)]
# setnames(param_df, old = c("Identifier", "UnitIdentifier"), new = c("Param", "Unit"))
select_col_2 <- c("Identifier", "UnitIdentifier")
new_name_2 <- c("Param", "Unit")
param_df <-
  lapply(param, "[", select_col_2) |>
  do.call(rbind, args = _) |>
  unlist() |>
  matrix(ncol = length(select_col_2)) |>
  unique() |>
  as.data.frame() |>
  `colnames<-`(new_name_2)


# ===================================
# --- Export the obtained information
# ===================================
path <- getwd()
path_info <- file.path(path, "info")
if (!dir.exists(path_info))
  dir.create(path_info)
# fwrite(plate_df, file.path(path_info, "plate_info.csv"))
# fwrite(param_df, file.path(path_info, "param_info.csv"))
write.csv(plate_df, file.path(path_info, "plate_info.csv"), row.names = FALSE)
write.csv(param_df, file.path(path_info, "param_info.csv"), row.names = FALSE)


message(
  "Time elapsed:\t",
  round(difftime(Sys.time(), time_start, units = "secs"), digits = 3),
  " seconds.\n\n"
)

File: scripts/python/0_update_param_site.py

import json
import time
from pathlib import Path

import urllib3

time_start = time.perf_counter()


# ==================================================
# --- Variables for connecting AQ (by `urllib3`) ---
# ==================================================
end_point = 'https://aquarius.orc.govt.nz/AQUARIUS/Publish/v2'
login = 'api-read:PR98U3SKOczINoPHo7WM'
http = urllib3.PoolManager()
hdr = urllib3.util.make_headers(basic_auth=login)


# =========================================================================
# --- 'GetLocationDescriptionList': plate numbers (ID) <-> Names (Site) ---
# =========================================================================
url_desc = f'{end_point}/GetLocationDescriptionList'
r_desc = http.request('GET', url_desc, headers=hdr)
plate_list = json.loads(r_desc.data.decode('utf-8')).get('LocationDescriptions')
plate_dict = {i.get('Identifier'): i.get('Name') for i in plate_list}


# ============================================
# --- 'GetParameterList': Unit_id <-> Unit ---
# ============================================
url_param = f'{end_point}/GetParameterList'
r_param = http.request('GET', url_param, headers=hdr)
param_list = json.loads(r_param.data.decode('utf-8')).get('Parameters')
param_dict = {i.get('Identifier'): i.get('UnitIdentifier') for i in param_list}


# ===================================
# --- Export the obtained information
# ===================================
if not (path_info := Path('info')).exists():
    path_info.mkdir()
with Path(path_info / 'plate_info.json').open('w') as fo:
    json.dump(plate_dict, fo, indent=4)
with Path(path_info / 'param_info.json').open('w') as fo:
    json.dump(param_dict, fo, indent=4)


print(f'\nTime elapsed:\t{(time.perf_counter() - time_start):.3f} seconds.')
